.golangci.example.yml

```diff
@@ -449,6 +449,12 @@ linters-settings:
     servingv1: knative.dev/serving/pkg/apis/serving/v1
     # using `autoscalingv1alpha1` alias for `knative.dev/serving/pkg/apis/autoscaling/v1alpha1` package
     autoscalingv1alpha1: knative.dev/serving/pkg/apis/autoscaling/v1alpha1
+  gomodreplace:
```

Comment:
I know my linter did not, so feel free to say we should add `gomodxxx`, but any chance we can expand the scope to restrict exclude and retract statements.

`retract` is new, and does generate stdout warnings when go-getting, but the code will build and run without them.

[`exclude`](https://github.com/golang/go/wiki/Modules#gomod) seems to be an escape hatch like replace, but for resolving conflicts or removing broken modules. it has a similar virality (must be in ALL main modules) to `replace`, so my gut say disallow.

Reply:
yes the scope can be expanded. I need to get more information about `exclude` directives.

about `retract` directives I don't see what rule do you want to apply, could you explain?

Reply:
the `exclude` are not transitive but I don't see how to handle that differently.

https://golang.org/ref/mod#go-mod-file-exclude

Reply:
for my repos, the lint for `exclude` and `replace` would be the same: disallow all. later we could add nuance to allow some `exclude` cases, but I have not seen it in the wild.

`retract` could be tricky. the ask is to "disallow any dependencies that themselves are retracted". now that I lay this out, it may be decent to do this as a separate linter, since the scope of "recurse through all dependencies and detect if their version is retracted" could get complex. your call. we may also run up against the fact that this resolution information only exists in `cmd/go/internal`.

another ask for `retract` is more a syntax thing: we should probably enforce that `modfile.File.Retract.Rationale != ""`.